backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -144,59 +144,60 @@ public CommCarePlatform getPlatform() {
      * Based on the current state of the session, determine what information is needed next to
      * proceed
      *
-     * @return 1 of the 4 STATE strings declared at the top of SessionFrame.java, or null if
+     * @return 1 of the 4 STATE strings declared at the top of SessionFrame, or null if
      * the session does not need anything else to proceed
      */
     public String getNeededData() {
-        // If we don't have a command yet, then need to get that first
-        if (this.getCommand() == null) {
+        if (currentCmd == null) {
             return SessionFrame.STATE_COMMAND_ID;
         }
 
-        Vector<Entry> possibleEntries = getEntriesForCommand(this.getCommand(), this.getData());
+        Vector<Entry> entries = getEntriesForCommand(currentCmd, collectedDatums);
+        String needDatum = getDataNeededByAllEntries(entries);
 
-        //Get data. Checking first to see if the relevant key is needed by all entries
+        if (needDatum != null) {
+            return needDatum;
+        } else if (entries.size() > 1 || !entries.elementAt(0).getCommandId().equals(currentCmd)) {
+            //the only other thing we can need is a form command. If there's
+            //still more than one applicable entry, we need to keep going
+            return SessionFrame.STATE_COMMAND_ID;
+        } else {
+            return null;
+        }
+    }
 
+    /**
+     * Checks that all entries have the same id for their first required data,
+     * and if so, returns the data's associated session state. Otherwise,
+     * returns null.
+     */
+    private String getDataNeededByAllEntries(Vector<Entry> entries) {
         String needDatum = null;
```

Comment:
I think renaming these to something like `datumNeededByAllEntriesSoFar` (and `idOfDatumNeededByAllEntriesSoFar`) would increase the readability of this by a lot